trieste/acquisition/function.py

```diff
@@ -744,10 +791,36 @@ def prepare_acquisition_function(
 
         feasible_query_points = tf.boolean_mask(objective_dataset.query_points, is_feasible)
         feasible_mean, _ = objective_model.predict(feasible_query_points)
+        self._update_expected_improvement_fn(objective_model, feasible_mean)
+
+        if self._constrained_improvement_fn is not None:
+            return self._constrained_improvement_fn
+
+        @tf.function
+        def constrained_function(x: TensorType) -> TensorType:
+            return cast(AcquisitionFunction, self._expected_improvement_fn)(x) * cast(
+                AcquisitionFunction, self._constraint_fn
+            )(x)
+
+        self._constrained_improvement_fn = constrained_function
+        return self._constrained_improvement_fn
+
+    def _update_expected_improvement_fn(
+        self, objective_model: ProbabilisticModel, feasible_mean: TensorType
+    ) -> None:
+        """
+        Set or update the unconstrained expected improvement function.
+
+        :param objective_model: The objective model.
+        :param feasible_mean: The mean of the feasible query points.
+        """
         eta = tf.reduce_min(feasible_mean, axis=0)
-        function = expected_improvement(objective_model, eta)
 
-        return lambda at: function(at) * constraint_fn(at)
+        if self._expected_improvement_fn is None:
+            self._expected_improvement_fn = expected_improvement(objective_model, eta)
+        else:
+            tf.debugging.Assert(isinstance(self._expected_improvement_fn, expected_improvement), [])
+            self._expected_improvement_fn.update(eta)  # type: ignore
```

Comment:
It there any reason why this is its own function?

Reply:
Yes, it allows ExpectedConstrainedHypervolumeImprovement to inherit from ExpectedConstrainedImprovement without having to duplicate lots of code. Also, it's now used both when preparing and updating the functions.